service/history/workflow/mutable_state_impl.go

```diff
@@ -2130,6 +2133,66 @@ func (ms *MutableStateImpl) addResetPointFromCompletion(
 	return true
 }
 
+// processBuildIdRedirect validates build id for the task being dispatched and applies possible redirect to
+// mutable state.
+// If a valid redirect is happening, assigned build id of the wf will be updated and all scheduled but not
+// started tasks will be rescheduled to be put on the matching queue of the right build id.
+// If the given versioning stamp and redirect info is not valid based on the wf's assigned build id
+// InvalidDispatchBuildId error will be returned.
+func (ms *MutableStateImpl) processBuildIdRedirect(
+	startingTaskScheduledEventId int64,
+	versioningStamp *commonpb.WorkerVersionStamp,
+	redirectInfo *taskqueue.BuildIdRedirectInfo,
+) error {
+	if !versioningStamp.GetUseVersioning() || versioningStamp.GetBuildId() != ms.GetAssignedBuildId() {
+		// dispatch build id is the same as wf assigned build id, hence noop.
+		return nil
+	}
+
+	if redirectInfo == nil || redirectInfo.GetAssignedBuildId() != ms.GetAssignedBuildId() {
+		// no redirect or a redirect based on a wrong assinged build id is reported. this must be a task
+		// backlogged on an old build id. rejecting this task, there should be another task scheduled on
+		// the right build id.
+		return serviceerrors.NewInvalidDispatchBuildId()
+	}
+
+	// redirect is valid, updating mutable state and re-scheduling pending tasks which are not started yet.
+	newBuildId := versioningStamp.GetBuildId()
+	err := ms.UpdateBuildIdAssignment(newBuildId)
+	if err != nil {
+		return err
+	}
+
+	if ms.HasPendingWorkflowTask() && !ms.HasStartedWorkflowTask() &&
```

Comment:
I don't have enough context to comment on this. I would get some extra eyes here.